pykokkos/lib/ufuncs.py

```diff
@@ -1381,6 +1389,270 @@ def exp2(view):
+    return out
+
+
+@pk.workunit
+def add_numm_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = viewA[tid] + viewB[0]
+
+@pk.workunit
+def add_num_2d_impl_1d_double(tid: int, viewA: pk.View2D[pk.double], viewB: pk.View1D[pk.double], out: pk.View2D[pk.double]):
+    for i in range(viewA.extent(1)):
+        out[tid][i] = viewA[tid][i] + viewB[i]
+
+def add_num(viewA, viewB):
+    if (not hasattr(viewB, '__iter__')):
+        view_temp = pk.View([1], pk.double)
+        view_temp[0] = viewB
+        viewB = view_temp
+
+    if len(viewA.shape) == 2:
+        out = pk.View(viewA.shape, pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            add_num_2d_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    elif str(viewA.dtype) in "DataType.double":
+        out = pk.View([viewA.shape[0]], pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            add_numm_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+def _get_shape(arr):
+    rows = len(arr)
+    cols = 1
+
+    if (rows > 0 and isinstance(arr[0], Sequence)):
+        cols = len(arr[0])
+    
+    return [rows, cols] if (cols > 1) else [rows]
+
+
+def array(view):
+    res = np.array(view)
+    view = pk.View(res.shape, pk.double)
+    view[:] = res
+
+    return view
+
+def transpose(view):
+    res = list(map(list, zip(*view)))
+    view = pk.View(_get_shape(res), pk.double)
+    view[:] = res
+
+    return view
+
+@pk.workunit
+def power_num_impl_2d_double(tid: int, viewA: pk.View2D[pk.double], viewB: pk.View1D[pk.double], out: pk.View2D[pk.double]):
+    for i in range(viewA.extent(1)):
+        out[tid][i] = pow(viewA[tid][i], viewB[0])
+
+
+def power_num(viewA, viewB):
+    if (not hasattr(viewB, '__iter__')):
+        view_temp = pk.View([1], pk.double)
+        view_temp[0] = viewB
+        viewB = view_temp
+
+    if str(viewA.dtype) == "DataType.double":
+        out = pk.View(viewA.shape, pk.double)
+        pk.parallel_for(
+            viewA.shape[0],
+            power_num_impl_2d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+
+@pk.workunit
+def index_impl_1d_double(tid: int, viewA: pk.View1D[pk.double], viewB: pk.View1D[pk.int32], out: pk.View1D[pk.double]):
+    out[tid] = viewA[viewB[tid]]
+
+def index(viewA, viewB):
+    if str(viewB.dtype) == "DataType.int32":
+        out = pk.View(viewB.shape, pk.double)
+        pk.parallel_for(
+            viewB.shape[0],
+            index_impl_1d_double,
+            viewA=viewA,
+            viewB=viewB,
+            out=out)
+    else:
+        raise RuntimeError("Incompatible Types")
+    return out
+
+
+def findMax(viewA):
```

Comment:
where do we use this?

Reply:
Used in the gaussian bayes algo to find max for pk.view of variances

Reply:
I think the original implementation called a `max()` method on the array itself, so we're changing the interface a bit here. I think this is fine for now, but can you rename this to `find_max` for now? Also add a TODO: comment at the point where this is used

Reply:
Oh yes, you are right. I will update the interface